### ISSUE TITLE
[API Proposal]: System.DataUri

### ISSUE THREAD

**Issue by zippy1981** [dotnet/runtime#85164]
### Background and motivation

A class for parsing, creating and dealing with DataUris built into the runtime.

I don't think it should inherit from `System.Uri`. It should have a constructor that takes a string and another that takes a Uri. There should be a `ToDataUri()` or `AsDataUri()` extension method for URIs.

Then methods and properties for:

* mime type
* Encoding
* getting the body as a 
  * string
  * byte array
  * stream

### API Proposal

```csharp
namespace System;

public class DataUri
{
    public static DataUri FromString(string);
    public static DataUri FromBytes(IEnumerable<byte>);

    public DataUri(string uri); // throws if invalid Uri
    
    public DataUri(Uri uri);

    public string MimeType { get; set; }

    public bool Base64Encode { get; set;  } 

    public ContentType GetContentType();  // This could throw exceptions so probably shouldn't be a property reads mime type

    public SetContentType(ContentType contentType); 

    public string GetContentsAsString(bool base64Encoded = false);

    public IEnumerable<byte> GetContentsAsBytes();

    public Stream GetContentsAsStream();

   public override string ToString(); //The Uri
}
```

### API Usage

```csharp

var dataUri = new DataUri('data:text/plain;,Hello+World!');
dataUri.Base64Encode = true;

Console.WriteLine(dataUri); // data:text/plain;base64,SGVsbG8gV29ybGQh
```

**Comment by svick** [comment]
I agree that the ability to work with `data` URIs is missing in .Net, though I have some suggestions about the API:

1. It shouldn't use `IEnumerable<byte>`, ever (which is horribly inefficient and not idiomatic). Instead a sequence of bytes should be represented as a `ReadOnlySpan<byte>`.
2. I find the distinction between `FromString` and the `string` constructor confusing. I think `FromString` and `FromBytes` should  be something like `FromContent(ReadOnlySpan<byte> content)` instead.
3. Consider having `Parse(ReadOnlySpan<byte>)` (and `TryParse`) instead of the `string` constructor.
4. It should use `Content`, not `Contents` (which is consistent with e.g. `HttpResponseMessage`).
5. I'm not sure I like the `base64Encoded` parameter on `GetContentsAsString`. Even though `data` URIs can use Base64 internally, I don't think that should be exposed when reading the data.
6. The API should be symmetric: if it allows reading the content as a `Stream`, it should probably also allow creating a `data` URI from a `Stream`.

**Comment by MihaZupan** [comment]
Triage:
- Moving to Future to gauge user interest in such functionality. Please upvote the top post if this impacts you to help us prioritize.
- We would make changes to the proposed API as suggested above when it comes to that

**Comment by dersia** [comment]
I really like the idea of having a DataUri class in the runtime, but I am not sure I like the proposed api. It is too expressive. There are many points that @svick made, that I agree with and I would go even further. When you create DataUri's from `ROS<byte>` or even `byte[]` it should be automatically base64 encoded. Also I don't think I will ever need the base64 string converted back, so I think once the DataUri is created it should only hold the DataUri represantation. So we should only be able to do a `ToString()` on it.
Also when creating a DataUri we should also have a second argument that is either a `string` or some sort of dictonary/set or `<string, string?>` where one can set all the attributes not just the contentType/mediaType. If I want to set the charset I should be allowed to. If the content that I have is already a base64encoded string, than I would `Add("base64", null)` to that dictionary.

The implentation should check if that attribute is not part of the dictionary and I create the DataUri from bytes it would add that attribute.

So the only data that the type would store is a string representation of the datauri.
